Add docs for pgmq (#29232)
supabase/supabase: apps/docs/content/guides/database/extensions/pgmq.mdx

---
id: 'pgmq'
title: 'pgmq: Queues'
description: 'pgmq: Managed queues in Postgres'
---

pgmq is a lightweight message queue built on Postgres.

## Features

- Lightweight - No background worker or external dependencies, just Postgres functions packaged in an extension
- "exactly once" delivery of messages to a consumer within a visibility timeout
- API parity with AWS SQS and RSMQ
- Messages stay in the queue until explicitly removed
- Messages can be archived, instead of deleted, for long-term retention and replayability

## Enable the extension

```sql

> create extension queue;

Reviewer: is this a separate extension that's a dependency or did you mean `create schema`?

Author: good catch, that was meant to be `create schema queue`;
ty

| delay      | integer | Time in seconds before the message becomes visible. Defaults to 0. |

Example:

{/* prettier-ignore */}
```sql
select * from pgmq.send('my_queue', '{"hello": "world"}');
 send 
------
    4
```

---

#### send_batch

Send 1 or more messages to a queue.

{/* prettier-ignore */}
```sql
pgmq.send_batch(
    queue_name text,
    msgs jsonb[],
    delay integer default 0
)
returns setof bigint
```

**Parameters:**

| Parameter  | Type    | Description                                                         |
| :--------- | :------ | :------------------------------------------------------------------ |
| queue_name | text    | The name of the queue                                               |
| msgs       | jsonb[] | Array of messages to send to the queue                              |
| delay      | integer | Time in seconds before the messages becomes visible. Defaults to 0. |

{/* prettier-ignore */}
```sql
select * from pgmq.send_batch(
    'my_queue',
    array[
      '{"hello": "world_0"}'::jsonb,
      '{"hello": "world_1"}'::jsonb
    ]
);
 send_batch 
------------
          1
          2
```

---

### Reading Messages

#### read

> Read 1 or more messages from a queue. The VT specifies the amount of time in seconds that the message will be invisible to other consumers after reading.

Reviewer: ```suggestion
Read 1 or more messages from a queue. The VT specifies the delay in seconds between reading and the message becoming invisible to other consumers.
```
